### PR TITLE
Second homework without tests

### PR DESCRIPTION
**Pull request by Radzivon**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1,67 +1,67 @@
...
-                  <target>1.8</target>
-              </configuration>
-          </plugin>
-          <plugin>
-              <groupId>org.apache.maven.plugins</groupId>
-              <artifactId>maven-surefire-plugin</artifactId>
-          </plugin>
-      </plugins>
-  </build>
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
```
Comment (CaH9I) [inline pom.xml:2]: Please, avoid changes like this. If you really wish to change formatting, do it out of scope of any pull request

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 2]
```diff
@@ -12,7 +9,7 @@
     private List<Product> productList;
 
     private ArrayListProductDao() {
-        productList = fillProductList();
+        productList = new ArrayList<>();
     }
 
     synchronized public static ArrayListProductDao getInstance() {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: not the best Singleton implementation

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 3]
```diff
@@ -21,39 +18,65 @@ synchronized public static ArrayListProductDao getInstance() {
...
-        result.add(new Product(11L, "simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
-        result.add(new Product(12L, "simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
-        result.add(new Product(13L, "simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
-        return result;
-    }
-
+  
     @Override
     public Product getProduct(Long id) {
         return productList.stream().filter(product -> product.getId()
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java:22]: not thread safe

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 4]
```diff
@@ -21,39 +18,65 @@ synchronized public static ArrayListProductDao getInstance() {
...
+                .orElseThrow(() -> new IllegalArgumentException("Product with " + id + " isn't"));
     }
 
-    @Override
-    public synchronized List<Product> findProducts() {
-        return productList.stream().
-                filter(product -> (product.getPrice() != null && product.getStock() > 0)).
-                collect(Collectors.toList());
+
+    public List<Product> findProducts(String query, String sort, String order) {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: the logic should be place in the intermediate (service) layer

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 5]
```diff
@@ -21,39 +18,65 @@ synchronized public static ArrayListProductDao getInstance() {
...
+                comparator = Comparator.comparing(Product::getId);
+            }
+        }
+        if ("desc".equals(order)) {
+            comparator = comparator.reversed();
+        }
+        return comparator;
+    }
+
+    private List<Product> search(String query) {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: the logic should be place in the intermediate (service) layer

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 6]
```diff
@@ -21,39 +18,65 @@ synchronized public static ArrayListProductDao getInstance() {
...
+                        .collect(Collectors.toList());
+            }
+        }
+        return productList.stream()
+                .filter(product -> (product.getPrice() != null && product.getStock() > 0))
+                .collect(Collectors.toList());
+
+    }
+
+    private Comparator<Product> getComparator(String sortBy, String order) {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: the logic should be place in the intermediate (service) layer

Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: is would be great, if you could implement something like `enum ProductSortOption implements Comparator<Product>`

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 4 above)
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: not thread safe

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 7]
```diff
@@ -21,39 +18,65 @@ synchronized public static ArrayListProductDao getInstance() {
...
     public Product getProduct(Long id) {
         return productList.stream().filter(product -> product.getId()
                 .equals(id))
                 .findFirst()
-                .orElseThrow(() -> new ProductNotFoundException("Product with " + id + " isn't"));
+                .orElseThrow(() -> new IllegalArgumentException("Product with " + id + " isn't"));
     }
 
-    @Override
-    public synchronized List<Product> findProducts() {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java:53]: leave this method and let `ProductService` use it

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 5 above)
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: the method doesn't work according to requirements

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 8]
```diff
@@ -21,39 +18,65 @@ synchronized public static ArrayListProductDao getInstance() {
...
-        return productList.stream().
-                filter(product -> (product.getPrice() != null && product.getStock() > 0)).
-                collect(Collectors.toList());
+
+    public List<Product> findProducts(String query, String sort, String order) {
+        if (query != null) {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Too many conditions. Don't use `if` for such a purpose.
Hint: set default value for method params

File: src/main/java/com/es/phoneshop/model/product/Product.java [hunk 9]
```diff
@@ -56,12 +58,14 @@ public void setDescription(String description) {
...
 
-    public BigDecimal getPrice() {
-        return price;
+    public List<ProductPrice> getPrice() {
+        return prices;
     }
 
-    public void setPrice(BigDecimal price) {
-        this.price = price;
+    public BigDecimal getActualPrice() { return prices.get(1).getPrice(); }
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/Product.java]: why `1`?

Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/Product.java]: rename it to `getPrice()`

File: src/main/java/com/es/phoneshop/model/product/Product.java [hunk 10]
```diff
@@ -56,12 +58,14 @@ public void setDescription(String description) {
         this.description = description;
     }
 
-    public BigDecimal getPrice() {
-        return price;
+    public List<ProductPrice> getPrice() {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/Product.java]: rename to `getPrices()`

File: src/main/java/com/es/phoneshop/model/product/Product.java [hunk 11]
```diff
@@ -56,12 +58,14 @@ public void setDescription(String description) {
...
+        return prices;
     }
 
-    public void setPrice(BigDecimal price) {
-        this.price = price;
+    public BigDecimal getActualPrice() { return prices.get(1).getPrice(); }
+
+    public void setPrice(List<ProductPrice> prices) {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/Product.java]: rename to `setPrices`

File: src/main/java/com/es/phoneshop/model/product/ProductDao.java [hunk 12]
```diff
@@ -3,8 +3,8 @@
 import java.util.List;
 
 public interface ProductDao {
-    Product getProduct(Long id);
-    List<Product> findProducts();
-    void save(Product product);
+    Product getProduct(Long id) throws IllegalArgumentException;
+    List<Product> findProducts(String query, String sort, String order);
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ProductDao.java]: move to service

File: src/main/java/com/es/phoneshop/model/product/ProductDao.java [hunk 13]
```diff
@@ -3,8 +3,8 @@
 import java.util.List;
 
 public interface ProductDao {
-    Product getProduct(Long id);
-    List<Product> findProducts();
-    void save(Product product);
+    Product getProduct(Long id) throws IllegalArgumentException;
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ProductDao.java]: why did you add `throws` statement?

File: src/main/java/com/es/phoneshop/model/product/ProductDao.java [hunk 14]
```diff
@@ -3,8 +3,8 @@
...
+    List<Product> findProducts(String query, String sort, String order);
+    void save(Product product) throws IllegalArgumentException;
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ProductDao.java]: why did you add `throws` statement?

File: src/main/java/com/es/phoneshop/model/product/ProductPrice.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
+package com.es.phoneshop.model.product;
+
+import java.math.BigDecimal;
+import java.util.Objects;
+
+public class ProductPrice {
+    String date;
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ProductPrice.java]: private

File: src/main/java/com/es/phoneshop/model/product/ProductPrice.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    BigDecimal price;
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ProductPrice.java]: private

File: src/main/java/com/es/phoneshop/model/product/ProductPrice.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    public BigDecimal getPrice() {
+        return price;
+    }
+
+    public void setPrice(BigDecimal price) {
+        this.price = price;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ProductPrice.java:32]: what if the same price value is used for different dates?

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+    private ProductDao productDao = ArrayListProductDao.getInstance();
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        request.setAttribute("product", productDao.getProduct(parseProductId(request)));
+        request.getRequestDispatcher("/WEB-INF/pages/product.jsp").forward(request, response);
+    }
+
+    private Long parseProductId(HttpServletRequest request) {
+        return Long.valueOf(request.getPathInfo().substring(1));
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:22]: what would happened if the result is not a number?

File: src/main/webapp/WEB-INF/web.xml [hunk 19]
```diff
@@ -3,16 +3,46 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>product</servlet-name>
+        <servlet-class>com.es.phoneshop.web.ProductDetailsPageServlet</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>product</servlet-name>
+        <url-pattern>/products/description/*</url-pattern>
```
Comment (CaH9I) [inline src/main/webapp/WEB-INF/web.xml]: use `/products/*` mapping

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class ProductDetailsPageServlet extends HttpServlet {
+    private ProductDao productDao = ArrayListProductDao.getInstance();
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        request.setAttribute("product", productDao.getProduct(parseProductId(request)));
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:17]: what if product with a given id doesn't exist?

File: src/main/webapp/WEB-INF/pages/product.jsp [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+<%@ taglib prefix="fmt" uri="http://java.sun.com/jsp/jstl/fmt" %>
+<%@ taglib prefix="tags" tagdir="/WEB-INF/tags" %>
+<jsp:useBean id="product" type="com.es.phoneshop.model.product.Product" scope="request"></jsp:useBean>
+<tags:master pageTitle="Products">
+    <img class="product-tile"
+         src="https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/${product.imageUrl}">
+    <br>
+    <p>${product.description}</p><br>
+    <fmt:formatNumber value="${product.price.get(0).price}" type="currency"
```
Comment (CaH9I) [inline src/main/webapp/WEB-INF/pages/product.jsp]: must be `${product.price}`

Comment (CaH9I) [inline src/main/webapp/WEB-INF/pages/product.jsp]: create custom `price.tag`

File: src/main/webapp/WEB-INF/pages/productList.jsp [hunk 22]
```diff
@@ -5,27 +5,40 @@
...
+        </tr>
+        </thead>
+        <c:forEach var="product" items="${products}">
+            <tr>
+                <td>
+                    <img class="product-tile"
+                         src="https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/${product.imageUrl}">
+                </td>
+                <td>
+                    <a href="${pageContext.servletContext.contextPath}/products/description/${product.id}">${product.description}</a>
```
Comment (CaH9I) [inline src/main/webapp/WEB-INF/pages/productList.jsp]: `<c:url>`

File: src/main/java/com/es/phoneshop/model/product/ProductDao.java [hunk 23]
```diff
@@ -4,7 +4,7 @@
 
 public interface ProductDao {
     Product getProduct(Long id);
-    List<Product> findProducts();
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ProductDao.java]: leave it

File: src/main/java/com/es/phoneshop/web/ProductDemodataServletContextListener.java [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+            fillProductList().forEach(productDao::save);
+        }
+    }
+
+    @Override
+    public void contextDestroyed(ServletContextEvent servletContextEvent) {
+
+    }
+
+    private synchronized List<Product> fillProductList() {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/web/ProductDemodataServletContextListener.java]: remove `synchronized`

File: src/main/webapp/WEB-INF/pages/productList.jsp (hunk 22 above)
Comment (Radzivon) [reply src/main/webapp/WEB-INF/pages/productList.jsp]: fix

File: src/main/webapp/WEB-INF/pages/product.jsp (hunk 21 above)
Comment (Radzivon) [reply src/main/webapp/WEB-INF/pages/product.jsp]: fix

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java (hunk 20 above)
Comment (Radzivon) [reply src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:17]: fix

File: src/main/webapp/WEB-INF/web.xml (hunk 19 above)
Comment (Radzivon) [reply src/main/webapp/WEB-INF/web.xml]: fix

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java (hunk 18 above)
Comment (Radzivon) [reply src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:22]: NumberFormatException

File: src/main/java/com/es/phoneshop/model/product/ProductPrice.java (hunk 17 above)
Comment (Radzivon) [reply src/main/java/com/es/phoneshop/model/product/ProductPrice.java:32]: fix

File: src/main/java/com/es/phoneshop/model/product/ProductPrice.java (hunk 16 above)
Comment (Radzivon) [reply src/main/java/com/es/phoneshop/model/product/ProductPrice.java]: fix

File: src/main/java/com/es/phoneshop/model/product/ProductPrice.java (hunk 15 above)
Comment (Radzivon) [reply src/main/java/com/es/phoneshop/model/product/ProductPrice.java]: fix

File: src/main/java/com/es/phoneshop/model/product/Product.java (hunk 11 above)
Comment (Radzivon) [reply src/main/java/com/es/phoneshop/model/product/Product.java]: fix

File: src/main/java/com/es/phoneshop/model/product/Product.java (hunk 10 above)
Comment (Radzivon) [reply src/main/java/com/es/phoneshop/model/product/Product.java]: fix

File: src/main/java/com/es/phoneshop/model/product/Product.java (hunk 9 above)
Comment (Radzivon) [reply src/main/java/com/es/phoneshop/model/product/Product.java]: fix

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 8 above)
Comment (Radzivon) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: ok

File: src/main/java/com/es/phoneshop/web/ProductDemodataServletContextListener.java (hunk 24 above)
Comment (Radzivon) [reply src/main/java/com/es/phoneshop/web/ProductDemodataServletContextListener.java]: fix

File: src/main/java/com/es/phoneshop/model/product/ProductDao.java (hunk 23 above)
Comment (Radzivon) [reply src/main/java/com/es/phoneshop/model/product/ProductDao.java]: fix

File: src/main/webapp/WEB-INF/pages/product.jsp (hunk 21 above)
Comment (Radzivon) [reply src/main/webapp/WEB-INF/pages/product.jsp]: fix
